### PR TITLE
IPv4/Single: Add a SocketID to a socket

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
In [issue #540](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/issues/543), [Jabob Carver](https://github.com/karver8) is asking for the possibility to assign a unique ID to each socket.

We found that a good idea and this PR adds it, depending on the new macro `ipconfigUSE_SetSocketID`.

In many applications, sockets are embedded in an object, a class. The "SetSocketID" can be a pointer to the instance of that class.
These will be the new functions:
~~~c
BaseType_t xSocketSetSocketID( const Socket_t xSocket, void * pvSocketID );
void * pvSocketGetSocketID( const ConstSocket_t xSocket );
~~~

Test Steps
-----------
~~~c
/* Read the current value of the SocketID. */
void * id1 = pvSocketGetSocketID( xSocket );

/* Set value of the SocketID. */
int rc = xSocketSetSocketID( xSocket, ( void * ) 0x12345678 );

/* Read the current value of the SocketID. */
void * id2 = pvSocketGetSocketID( xSocket );

FreeRTOS_printf( ( "SocketSetSocketID %p -> %p rc = %d", id1, id2, ( int ) rc ) );
~~~

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/include/FreeRTOSIPConfigDefaults.h [hunk 1]
```diff
@@ -1068,4 +1068,12 @@
     #define ipconfigPROCESS_CUSTOM_ETHERNET_FRAMES    0
 #endif
 
+/* This option adds the possibility to have a user-ID attached to a socket.
+ * The type of this ID is a void *.  Both UDP and TCP sockets have
+ * this ID. It has a default value of NULL.
+ */
+#ifndef ipconfigUSE_SetSocketID
```
Comment (AniruddhaKanhere) [inline source/include/FreeRTOSIPConfigDefaults.h:1075]: Do we really need one more config?

File: source/FreeRTOS_Sockets.c [hunk 2]
```diff
@@ -4780,6 +4780,55 @@ BaseType_t xSocketValid( const ConstSocket_t xSocket )
...
 
+#if ( ipconfigUSE_SetSocketID != 0 )
+
+/**
+ * @brief Set the value of the SocketID of a socket.
+ * @param[in] xSocket: The socket whose ID should be set.
+ * @param[in] pvSocketID: The new value for the SocketID.
+ * @return Zero if the socket was valid, otherwise -EINVAL.
+ */
+    BaseType_t xSocketSetSocketID( const Socket_t xSocket,
```
Comment (AniruddhaKanhere) [inline source/FreeRTOS_Sockets.c:4791]: Seems like this is an API. Why not stick to the convention of starting the name with `FreeRTOS_`?
Something like `FreeRTOS_SetSocketID`?

File: source/FreeRTOS_Sockets.c [hunk 3]
```diff
@@ -4780,6 +4780,55 @@ BaseType_t xSocketValid( const ConstSocket_t xSocket )
...
+#if ( ipconfigUSE_SetSocketID != 0 )
+
+/**
+ * @brief Retrieve the SocketID that is associated with a socket.
+ * @param[in] xSocket: The socket whose ID should be returned.
+ * @return The current value of pvSocketID, or NULL in case
+ *         the socket pointer is not valid or when the ID was not
+ *         yet set.
+ */
+    void * pvSocketGetSocketID( const ConstSocket_t xSocket )
```
Comment (AniruddhaKanhere) [inline source/FreeRTOS_Sockets.c:4817]: Here too!

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]

File: source/include/FreeRTOSIPConfigDefaults.h (hunk 1 above)
Comment (htibosch) [reply source/include/FreeRTOSIPConfigDefaults.h:1075]: Yes why not add it unconditionally. Linkers are smart enough to eliminate dead code. Only MISRA won't like it.
So I also vote to leave out the macro.

File: source/FreeRTOS_Sockets.c (hunk 2 above)
Comment (htibosch) [reply source/FreeRTOS_Sockets.c:4791]: A disadvantage of `FreeRTOS_SetSocketID` would be that the name doesn't carry the type prefixes, like "x" and "pv".

File: source/include/FreeRTOSIPConfigDefaults.h (hunk 1 above)
Comment (htibosch) [reply source/include/FreeRTOSIPConfigDefaults.h:1075]: > Yes why not add it unconditionally. Linkers are smart enough to eliminate dead code. Only MISRA won't like it. So I also vote to leave out the macro.

There is only one thing: all sockets will grown 4 of 8 bytes in size:
~~~
#if ( ipconfigUSE_SetSocketID != 0 )
    void * pvSocketID;
#endif
~~~
when the ID pointer is declared unconditionally.

Comment (AniruddhaKanhere) [reply source/include/FreeRTOSIPConfigDefaults.h:1075]: Yes, that is the cost of cleanliness in the code I suppose.
We already have far too many configuration options in my opinion. And the list just keeps growing.
I don't mind adding configs for things which affect the performance or certain characteristics (like timeouts etc) of the stack. But this seems like a bit of an overkill to me.
What do you think @htibosch?

**Comment by htibosch** [comment]
About "[Removed the 'ipconfigUSE_SetSocketID' option](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/546/commits/1e5ac852614ff648e02aec056208dfac43c599be)":
We decided to add the Socket-ID functions unconditionally.

**Review by AniruddhaKanhere — APPROVED** [review]

File: test/unit-test/FreeRTOS_Sockets/FreeRTOS_Sockets_GenericAPI_utest.c [hunk 4]
```diff
@@ -2903,6 +2903,89 @@ void test_FreeRTOS_maywrite_HappyPath( void )
...
+ */
+void test_xSocketSetSocketID_ValidSocket( void )
+{
+    BaseType_t xReturn;
+    FreeRTOS_Socket_t xSocket;
+    BaseType_t AnchorVariable;
+
+    memset( &xSocket, 0, sizeof( xSocket ) );
+
+    xReturn = xSocketSetSocketID( &xSocket, &AnchorVariable );
```
Comment (shubnil) [inline test/unit-test/FreeRTOS_Sockets/FreeRTOS_Sockets_GenericAPI_utest.c:2941]: What will be the value of the anchor variable here.

File: source/FreeRTOS_Sockets.c [hunk 5]
```diff
@@ -4797,6 +4797,49 @@ BaseType_t xSocketValid( const ConstSocket_t xSocket )
...
+ * @param[in] xSocket: The socket whose ID should be set.
+ * @param[in] pvSocketID: The new value for the SocketID.
+ * @return Zero if the socket was valid, otherwise -EINVAL.
+ */
+BaseType_t xSocketSetSocketID( const Socket_t xSocket,
+                               void * pvSocketID )
+{
+    FreeRTOS_Socket_t * pxSocket = ( FreeRTOS_Socket_t * ) xSocket;
+    BaseType_t xReturn = -pdFREERTOS_ERRNO_EINVAL;
+
```
Comment (shubnil) [inline source/FreeRTOS_Sockets.c:4811]: I have a query here. How are we getting the value of the socket_id. I mean, is it left to the test application to assign a value or will the TCP stack maintain a sequence.

**Review by shubnil — COMMENTED** [review]
Hi, Overall I wanted to understand more on how we are generating the socket-id. Will this be assigned by the test application or will the TCP stack has a way to maintain and assign a socket id.

**Comment by htibosch** [comment]
@shubnil: I will be one of the happy user of this new feature!

My applications encapsulates +TCP sockets into an object. This can be a struct or even a C++ class.

And as you know, the FreeRTOS+TCP library offers several different callbacks ( application hooks ) that are triggered on certain events: on connect, on read, on write etc.
Before this PR, I had to maintain a list of sockets, in order to be able to translate a socket pointer to its encapsulating object.

After this PR, that administration isn't necessary anymore.

The FreeRTOS+TCP library doesn't care about this pointer, it is just stores and reveals it.

> How are we getting the value of the socket_id. I mean, is it left to the test application to assign a value or will the TCP stack maintain a sequence.

As you will understand by now, the application builder will manage these pointers.

>> xReturn = xSocketSetSocketID( &xSocket, &AnchorVariable );
> What will be the value of the anchor variable here.

The new feature is storing a pointer. The value of the anchor value will probably an index or a pointer to an instance of a C++ class.

Does that make sense?

**Review by shubnil — APPROVED** [review]
Thanks, This makes sense.

**Pull request merged**